src/sentry/api/endpoints/source_map_debug.py

```diff
@@ -191,9 +189,92 @@ def _find_matches(self, release_artifacts, unified_path):
             artifact
             for artifact in release_artifacts
             if artifact.name.endswith(unified_path.split("/")[-1])
+            if artifact.name.endswith(unified_path.split("/")[-1])
         ]
         return full_matches, partial_matches
 
+    def _extract_release(self, event, project):
+        release_version = event.get_tag("sentry:release")
+
+        if not release_version:
+            return self._create_response(issue=SourceMapProcessingIssue.MISSING_RELEASE)
+        try:
+            release = Release.objects.get(
+                organization=project.organization, version=release_version
+            )
+        except Release.DoesNotExist:
+            return self._create_response(issue=SourceMapProcessingIssue.MISSING_RELEASE)
+
+        return release
+
+    def _verify_dist_matches(self, release, event, artifact, filename):
+        try:
+            dist = Distribution.objects.get(release=release, name=event.dist)
+        except Distribution.DoesNotExist:
+            return self._create_response(
+                issue=SourceMapProcessingIssue.DIST_MISMATCH,
+                data={"eventDist": event.dist, "fileName": filename},
+            )
+
+        if artifact.dist_id != dist.id:
+            return self._create_response(
+                issue=SourceMapProcessingIssue.DIST_MISMATCH,
+                data={"eventDist": dist.id, "artifactDist": artifact.dist_id, "fileName": filename},
+            )
+        return dist
+
+    def _find_matching_artifact(self, release_artifacts, urlparts, abs_path, filename):
+        unified_path = self._unify_url(urlparts)
+        full_matches, partial_matches = self._find_matches(release_artifacts, unified_path)
+
+        artifact_names = [artifact.name for artifact in release_artifacts]
+
+        if len(full_matches) == 0:
+            if len(partial_matches) > 0:
```

Comment:
I'm curious if you have more thoughts on the best way to implement this. Will DM for more info.

Reply:
I'd do matching in reverse now. From the full path, down to just filename, instead of reducing it to filename only.

Let's assume we have files like so (in this specific order):

```
~/v1/assets/header/bundle.js
~/v1/assets/footer/bundle.js
```

and we get event with `~/v1/scripts/footer/bundle.js` as its `abs_path`.

If we do partial matching as its implemented now, we'll report `~/v1/assets/header/bundle.js` as it matches `bundle.js` and is first match.

If we do it in reverse, trimming each segment on every iteration, we will do it correctly like so:
- do we match `v1/scripts/footer/bundle.js`? no, lets trim front path
- do we match `scripts/footer/bundle.js`? no, lets trim front path
- do we match `footer/bundle.js`? yes, let's report `~/v1/assets/footer/bundle.js` as possible match?

And we can explain to people somewhere, that wat dictates the path here is where the file was 
found from the root (using `includes` config options) and `urlPrefix`.